src/rsz/src/RepairSetup.cc

```diff
@@ -108,58 +108,123 @@ RepairSetup::init()
 {
   init();
+  constexpr int digits = 3;
   inserted_buffer_count_ = 0;
   resize_count_ = 0;
   resizer_->buffer_moved_into_core_ = false;
 
-  Slack worst_slack;
-  Vertex *worst_vertex;
-  sta_->worstSlack(max_, worst_slack, worst_vertex);
-  debugPrint(logger_, RSZ, "repair_setup", 1, "worst_slack = {}",
-             delayAsString(worst_slack, sta_, 3));
-  Slack prev_worst_slack = -INF;
-  int pass = 1;
-  int decreasing_slack_passes = 0;
+  // Sort failing endpoints by slack.
+  VertexSet *endpoints = sta_->endpoints();
+  VertexSeq violating_ends;
+  for (Vertex *end : *endpoints) {
+    Slack end_slack = sta_->vertexSlack(end, max_);
+    if (end_slack < setup_slack_margin)
+      violating_ends.push_back(end);
+  }
+  sort(violating_ends, [=](Vertex *end1, Vertex *end2) {
+    return sta_->vertexSlack(end1, max_) < sta_->vertexSlack(end2, max_);
+  });
+  debugPrint(logger_, RSZ, "repair_setup", 1, "Violating endpoints {}/{} {}%",
+             violating_ends.size(),
+             endpoints->size(),
+             int(violating_ends.size() / double(endpoints->size()) * 100));
+
+  int end_index = 0;
+  int max_end_count = violating_ends.size() * repair_tns_end_percent;
+  // Always repair the worst endpoint, even if tns percent is zero.
+  max_end_count = max(max_end_count, 1);
   resizer_->incrementalParasiticsBegin();
-  while (fuzzyLess(worst_slack, setup_slack_margin)
-         && pass <= max_passes) {
-    PathRef worst_path = sta_->vertexWorstSlackPath(worst_vertex, max_);
-    bool changed = repairSetup(worst_path, worst_slack);
+  for (Vertex *end : violating_ends) {
     resizer_->updateParasitics();
     sta_->findRequireds();
+    Slack end_slack = sta_->vertexSlack(end, max_);
+    Slack worst_slack;
+    Vertex *worst_vertex;
     sta_->worstSlack(max_, worst_slack, worst_vertex);
-    bool decreasing_slack = fuzzyLessEqual(worst_slack, prev_worst_slack);
-    debugPrint(logger_, RSZ, "repair_setup", 1, "pass {} worst_slack = {} {}",
-               pass,
-               delayAsString(worst_slack, sta_, 3),
-               decreasing_slack ? "v" : "^");
-    if (decreasing_slack) {
-      // Allow slack to increase to get out of local minima.
-      // Do not update prev_worst_slack so it saves the high water mark.
-      decreasing_slack_passes++;
-      if (!changed
-          || decreasing_slack_passes > repair_setup_decreasing_slack_passes_allowed_) {
-        // Undo changes that reduced slack.
+    debugPrint(logger_, RSZ, "repair_setup", 1, "{} slack = {} worst_slack = {}",
+               end->name(network_),
+               delayAsString(end_slack, sta_, digits),
+               delayAsString(worst_slack, sta_, digits));
+    end_index++;
+    if (end_index > max_end_count)
+      break;
+    Slack prev_end_slack = end_slack;
+    Slack prev_worst_slack = worst_slack;
+    int pass = 1;
+    int decreasing_slack_passes = 0;
+    resizer_->journalBegin();
+    while (pass <= max_passes) {
```

Comment:
what happens if there are more than 10,000 violating paths? The default for maxpasses is 10,000 per the tcl interface and this limit seems okay on smaller designs but large systems that is really easy to run into.

Reply:
I suspect the run time would be ridiculous repairing 10000 endpoints (not really "paths" because any endpoing can have an exponential number of paths to it). I tried a number of designs in the CI (after adjusting the clock so there was negative slack because most of them pass timing). There is also a -max_passes arg to repair_timing....